---
layout: issue
title: "Zend_Controller_Request_Http::setRequestUri() broken for certain requests"
id: ZF-6271
---

ZF-6271: Zend\_Controller\_Request\_Http::setRequestUri() broken for certain requests
-------------------------------------------------------------------------------------

 Issue Type: Bug Created: 2009-04-11T08:53:03.000+0000 Last Updated: 2009-09-01T15:13:50.000+0000 Status: Resolved Fix version(s): 
 Reporter:  Alex Adriaanse (alexadriaanse)  Assignee:  Matthew Weier O'Phinney (matthew)  Tags: - Zend\_Controller
 
 Related issues: 
 Attachments: - [Zend\_Controller\_Request\_Http\_setRequestUri\_fix.patch](/issues/secure/attachment/11868/Zend_Controller_Request_Http_setRequestUri_fix.patch)
 
### Description

Zend\_Controller\_Request::setRequestUri() is broken for URLs that happen to have the host also embedded inside the query. For example, <http://example.com/?redirect=http://example.com>. When setRequestUri() is called in such scenarios, and $\_SERVER['REQUEST\_URI'] contains a relative URL (e.g. /?redirect=<http://example.com>), the code mistakenly treats it as an absolute URL (since the value of $\_SERVER['HTTP\_HOST'] is embedded inside the query), and calls parse\_url() on this relative URL. parse\_url() is not designed to work on relative URLs (see <http://bugs.php.net/bug.php?id=43411>), so you get warnings such as these:

 
    <pre class="literal">
    PHP Warning:  parse_url(/?redirect=<a href="http://example.com">http://example.com</a>): Unable to parse URL in .../library/Zend/Controller/Request/Http.php on line 394
    PHP Warning:  parse_url(/?redirect=<a href="http://example.com">http://example.com</a>): Unable to parse URL in .../library/Zend/Controller/Request/Http.php on line 395


I am attaching a patch that fixes this by changing how absolute URLs are recognized.

 

 

### Comments

Posted by Ondrej Balaz (balazo) on 2009-04-24T05:34:06.000+0000

Patch fixes same issue for me.

 

 

Posted by Matthew Weier O'Phinney (matthew) on 2009-04-24T06:07:18.000+0000

I believe this is already fixed on current trunk; could one of you test with either current svn trunk or the 1.8 beta release and verify?

 

 

Posted by Micha? Zieli?ski (zielun) on 2009-06-08T00:13:13.000+0000

I still got this issue with 1.7.8 I`m prepering to migrate to 1.8.x within next 2 weeks so I`ll test this out. Anyway, I`ve got quite a lot of this type of warning in my logs :/

 

 

Posted by Alex Adriaanse (alexadriaanse) on 2009-09-01T15:13:49.000+0000

I just tested against 1.8.4PL1 and the problem has been solved. Thanks!

 

 